seDataDevices.py

```diff
@@ -0,0 +1,731 @@
+            printPtr = 0
+            while len(data) - printPtr >= lineSize:
+                hexLines.append(hexLine(data[printPtr:printPtr + lineSize]))
+                printPtr += lineSize
+            if printPtr < len(data):
+                hexLines.append(hexLine(data[printPtr:]))
+        return ' | '.join(hexLines)
+
+    @staticmethod
+    def unhexData(hexData):
+        """
+        Convert a hex representation into a string of bytes, ie the inverse of hexData.
+
+        :param hexData: String, containing a (must be valid) hex representation of some bytes.
+        :return: The string of bytes corresponding to hexData.
+        """
+        return binascii.unhexlify(hexData.replace("|","").replace(" ",""))
+
+    def __init__(self, data, explorer=False):
+        self.parseDevTable(data)
+        self.setDerivationDefaults()
+        self.codeDerivations()
+        self.checkHypotheses()
+
+    def parseDevTable(self, data):
+
+        dataPtr = 0
+        devHdrLen = 8
+        # device header
+        (seType, seId, devLen) = struct.unpack("<HLH", data[dataPtr:dataPtr + devHdrLen])
+        dataPtr += devHdrLen
+
+        # For (almost) all subclasses, _devType will already have this value.
+        # This is necessary only when a default catchall parse is happening, because a specific parser for seType has
+        # not been defined.
+        self._devType = '{}_{:#06x}'.format(self._devName, seType)
+        # Store seId as an attribute for later use as part of the standard "dictionary of dictionaries" wrapper when
+        # a parsed instance is converted to json.
+        self._seId = self.parseId(seId)
+
+        self.update({'seType': '{:#06x}'.format(seType),
+                     'seId': self._seId,
+                     'devLen': devLen,
+                     'devType' : self._devType}
+                    )
+
+        if self.defnLen > devLen :
+            raise ValueError('You have defined more bytes, {}, than the message contains, {}'.format(self.defnLen, devLen))
+        elif self.defnLen < devLen :
+            # By default, convert any remaining undefined bytes to their representation as a hexadecimal string.
+            self._defn.append([devLen - self.defnLen, 'hex', "Undeciphered_data", self.hexData, True, "Unknown as yet" ])
+
+        for paramLen, paramInFmt, paramName, outFormatFn, out, comment in self._defn:
+            # Extract the field
+            if paramInFmt == 'hex':
+                self[paramName] = data[dataPtr: dataPtr + paramLen]
+            # Test for nan float values and set them to 'nan'
+            # On my machine, '7fffffff' is unpacked to nan for both big endian and little endian formats.
+            # 'ffff7fff' also unpacks as nan for big endian, BUT as -3.402...*10**38 for little endian.
+            # I suspect a bug somewhere, but in the meantime just check the bytes
+            elif paramInFmt == 'f' and (data[dataPtr: dataPtr + paramLen] == '\xff\xff\x7f\xff'):
```

Comment:
I just now realized that @Geoff99 is saying that '\xff\xff\x7f\xff' is a NaN represented in the opposite endianness from the rest of the data in the messages.  Probably there is some legacy reason for this, but that's what we have to work with.